components/locale_core/src/data.rs

```diff
@@ -0,0 +1,350 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::extensions::unicode as unicode_ext;
+use crate::subtags::{Language, Region, Script, Subtag, Variant};
+use crate::{LanguageIdentifier, Locale, ParseError};
+use core::cmp::Ordering;
+use core::default::Default;
+use core::fmt;
+use core::hash::Hash;
+use core::str::FromStr;
+
+/// A locale type optimized for use in fallbacking and the ICU4X data pipeline.
+///
+/// [`DataLocale`] contains less functionality than [`Locale`] but more than
+/// [`LanguageIdentifier`] for better size and performance while still meeting
+/// the needs of the ICU4X data pipeline.
+///
+/// You can create a [`DataLocale`] from a borrowed [`Locale`], which is more
+/// efficient than cloning the [`Locale`], but less efficient than converting an owned
+/// [`Locale`]:
+///
+/// ```
+/// use icu_locale_core::locale;
+/// use icu_provider::DataLocale;
+///
+/// let locale1 = locale!("en-u-ca-buddhist");
+/// let data_locale = DataLocale::from(&locale1);
+/// ```
+///
+/// [`DataLocale`] only supports `-u-sd` keywords, to reflect the current state of CLDR data
+/// lookup and fallback. This may change in the future.
+///
+/// ```
+/// use icu_locale_core::{locale, Locale};
+/// use icu_provider::DataLocale;
+///
+/// let locale = "hi-IN-t-en-h0-hybrid-u-attr-ca-buddhist-sd-inas"
+///     .parse::<Locale>()
+///     .unwrap();
+///
+/// assert_eq!(
+///     DataLocale::from(locale),
+///     DataLocale::from(locale!("hi-IN-u-sd-inas"))
+/// );
+/// ```
+#[derive(Clone, Default, PartialEq, Hash, Eq)]
+#[allow(clippy::exhaustive_structs)] // See #3632
+pub struct DataLocale {
+    /// Language subtag
+    pub language: Language,
+    /// Script subtag
+    pub script: Option<Script>,
+    /// Region subtag
+    pub region: Option<Region>,
+    /// Variant subtag
+    pub variant: Option<Variant>,
+    /// Subivision (-u-sd-) subtag
+    pub subdivision: Option<Subtag>,
+}
+
+impl DataLocale {
+    #[doc(hidden)]
```

Comment:
I wasn't sure what to do about this one: it's used in icu_provider for comparing, but I don't want to expose DataLocale internals. 

My suggestion: we land this as is and we discuss how to handle DataLocale's internal-private in the upcoming PR. This PR is more about changing location and other refactors, and is big enough as it is.

Reply:
I think we should fully align the `Locale`, `LanguageIdentifier`, and `DataLocale` APIs anyway.